### PR TITLE
feat: Add manual dataloader reloading feature (Closes #21448)

### LINKED ISSUE #21448: Manual `Dataloader` reloading during training
### Description & Motivation

Currently, PyTorch Lightning assumes that `Dataset` and `DataLoader` parameters (e.g., sequence length, unrolling steps $N→N+m$, or sampling ratios) remain static throughout the lifecycle of a `fit()` call (unless you set a static policy using `reload_dataloader_every_n_epochs` in `Trainer`). When a user needs to adjust these parameters dynamically based on real-time metrics (such as curriculum learning or adaptive unrolling), they are forced into two suboptimal patterns:

- Manual Script Restarts: Stopping execution, saving a checkpoint, updating the configuration, and resuming. This introduces significant overhead and breaks the continuity of the training loop.
- Dataset Over-engineering: Building complex, stateful Dataset objects that internally track metrics and handle logic.

### Pitch

I propose introducing a formal mechanism to trigger in-place `DataLoader` reconfiguration without exiting the `fit()` loop. This would likely manifest as a new hook or an enhancement to the existing `reload_dataloaders_every_n_epochs` functionality, allowing for:

- Metric-driven triggers: Re-initializing the `DataLoader` when a specific monitor value hits a threshold.
- Parameter Injection: A way to pass updated hyper-parameters from the LightningModule or a Callback directly into the datamodule during the reload process.

### Alternatives

_No response_

### Additional context

Here a snapshot of how a callback could be implemented if the requested feature will be accepted

```python
def on_train_batch_end(self, outputs, batch, batch_idx):
    if self.trainer.callback_metrics['train_loss'] < 0.5:  # this is just an example (more complex logic will be fine)
        # Trigger a reload with updated sequence length
        self.trainer.datamodule.unroll_steps += 1
        self.trainer.manual_dataloader_reload()
```

I already have a draft code for this improvement and it will be a pleasure for me to open a PR is the feature will be of your interest

cc @lantiga

### PR DESCRIPTION
**Pull request by arrdel**
This PR implements manual dataloader reloading for dynamic DataLoader reconfiguration during training without exiting the fit loop. Addresses issue #21448.

----
📚 Documentation preview 📚: https://pytorch-lightning--21473.org.readthedocs.build/en/21473/

### CODE REVIEW

**Comment by arrdel** [comment]
Hi @FilippoOlivo, thanks for the suggestion! That is an interesting approach to add more flexibility. The current implementation already provides manual control through the trainer.reload_dataloaders(train=True/False, val=True/False) method. But I see your point about adding a force parameter within setup_data() itself. This could allow for more granular control directly inside the loop logic. We need to ensure this does not conflict with the existing _should_reload_train_dl / _should_reload_val_dl flags that currently manage the reload state. Thanks again for the thoughtful suggestion!

**Comment by FilippoOlivo** [comment]
> Hi @FilippoOlivo, thanks for the suggestion! That is an interesting approach to add more flexibility. The current implementation already provides manual control through the trainer.reload_dataloaders(train=True/False, val=True/False) method. But I see your point about adding a force parameter within setup_data() itself. This could allow for more granular control directly inside the loop logic. We need to ensure this does not conflict with the existing _should_reload_train_dl / _should_reload_val_dl flags that currently manage the reload state. Thanks again for the thoughtful suggestion!

Hi, if you prefer I can try to implement it. Let me know!

**Comment by arrdel** [comment]
Rebased on latest upstream/master to fix CI failures. The branch was 55 commits behind, which was causing the test failures. CI should pass now after the rebase.
